tsdb/querier.go

```diff
@@ -584,43 +584,47 @@ func (p *populateWithDelGenericSeriesIterator) reset(blockID ulid.ULID, cr Chunk
 	p.currChkMeta = chunks.Meta{}
 }
 
-func (p *populateWithDelGenericSeriesIterator) next() bool {
+// If copyHeadChunk is true, then the head chunk (i.e. the in-memory chunk of the TSDB)
+// is deep copied to avoid races between reads and copying chunk bytes.
+// However, if the deletion intervals overlaps with the head chunk, then the head chunk is
+// not copied irrespective of copyHeadChunk because it will be re-encoded later anyway.
```

Comment:
Note to self: the re-encoding happens in `populateWithDelChunkSeriesIterator.Next()`.
`populateWithDelGenericSeriesIterator` is also used by `populateWithDelSeriesIterator`, which as far as I can see does not have the re-encoding behaviour.